qemu/tests/hugepage_specify_node.py

```diff
@@ -0,0 +1,83 @@
+import logging
+import math
+
+from avocado.utils import memory
+
+from virttest import env_process
+from virttest import error_context
+from virttest import utils_misc
+
+from virttest.utils_numeric import normalize_data_size
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Qemu allocate hugepage from specify node.
+    Steps:
+    1) Setup total of 4G mem hugepages for specify node.
+    2) Setup total of 1G mem hugepages for idle node.
+    3) Mount this hugepage to /mnt/kvm_hugepage.
+    4) Boot guest only allocate hugepage from specify node.
+    5) Check the hugepage used from every node.
+    :params test: QEMU test object.
+    :params params: Dictionary with the test parameters.
+    :params env: Dictionary with test environment.
+    """
+    memory.drop_caches()
+    hugepage_size = memory.get_huge_page_size()
+    mem_size = int(normalize_data_size("%sM" % params["mem"], "K"))
+    idle_node_mem = int(normalize_data_size("%sM" % params["idle_node_mem"], "K"))
+
+    error_context.context("Get host numa topological structure.", logging.info)
+    host_numa_node = utils_misc.NumaInfo()
+    node_list = host_numa_node.get_online_nodes_withmem()
+    idle_node_list = node_list.copy()
+    node_meminfo = host_numa_node.get_all_node_meminfo()
+
+    for node_id in node_list:
+        error_context.base_context("Check preprocess HugePages Free on host "
+                                   "numa node %s." % node_id, logging.info)
+        node_memfree = int(node_meminfo[node_id]["MemFree"])
+        if node_memfree < idle_node_mem:
+            idle_node_list.remove(node_id)
+        if node_memfree < mem_size:
+            node_list.remove(node_id)
+
+    if len(idle_node_list) < 2:
```

Comment:
> If node_list is empty, the framework will cancel this test.

 Would you please tell where framework cancel the test when node_list is empty?

Reply:
def set_node_num_huge_pages(self, num, node, pagesize):
......
        if obj.sys_fs_value < int(num):
            raise ValueError("Cannot set %s hugepages on node %s, please check"
                             " if the node has enough memory" % (num, node))

Reply:
If node_list is empty, the code after L51 won't be executed at all. Then this function won't run.

Reply:
correct